[RFC] Built-in LSP Support (#10222)
neovim/neovim: runtime/doc/lsp.txt

    startup activities whenever the |FileType| event is triggered for {ftype}.

    Parameters:~
        {ftype}: A string or list of strings of filetypes to associate with
        this server.

        {cmd}: See |jobstart()| for information on cmd.

        {configuration} (dictionary): Optional configuration for a server.
            `name`      (string):
              The name of the server

            `callbacks` (dict):
              A dictionary of callbacks to configure the server.

              Available callbacks are:

              `root_uri`: function(server) -> string.
                This function returns a string to populate the `root_uri`
                parameter in the initialize request for a server.

                Example:~
>
  " Python example
  call lsp#server_config#add('python', 'pyls')

  " Typescript configuration
  call lsp#server_config#add('typescript', {
      \ 'name': 'ts', ....
      \ 'callbacks': {
        \ 'root_uri': { server -> RootFinderFromPlugin() }
      \ }})

================================================================================
                                                                  *lsp-request*
Requests

> To make a request of the language server client running in Nvim, the following

Reviewer: ```suggestion
To make a request to the language server client running in Nvim, the following
```
?

            with reasonable default values.
                Example (with 'textDocument/hover'):
                    Passed in:
                        `{}` or |v:null|

                    Sent:
>
                        {
                            'textDocument': '/home/user/test/example.lua',
                            'position': { 'line': 13, 'character': 17 }
                        }
<
                Example (with 'textDocument/rename'):
                    Passed in:
>
                        {
                            'newName': 'exampleRename'
                        }
<
                    Sent:
>
                        {
                            'textDocument': '/home/user/test/rename.lua',
                            'position': { 'line': 16, 'character': 8 }
                            'newName': 'exampleRename'
                        }
<

            [callback]: An optional callback that will be executed
            once the response has been received. If this is passed and not
            |v:null|, then the |lsp-built-in-callback| will not be executed for
            this request. If you do not set |built-in-callbacks| or this request does not have a

>             a specify filetype callback, then no callback will be executed upon receiving the request.

Reviewer: ```suggestion
            specific filetype callback, then no callback will be executed upon receiving the request.
```

            The callback must be a string with the name of the vimscript
            function to call. The signature must be:

                `func_name({success}, {data})`

                where {success} is a boolean where true means successful request
                and false means the request failed.

                The parameter {data} is a dictionary containing the possible

>                 response data (or error) data as specified by the language

Reviewer: ```suggestion
                response data (or error) as specified by the language
```

                server protocol.

>                 The callback will be executed from lua running:

Reviewer: ```suggestion
                The callback will be executed from Lua running:
```

                    `vim.api.nvim_call_function('func_name', { success, data })`

            [filetype]: The filetype associated with the server.
            This will default to the current |filetype| if nothing is passed.

            Configuration for the default server can be found at
            |lsp-configuration|.

                                                          *lsp#request_async()*
lsp#request_async({method} [, {arguments} [, {callback} [, {filetype}]]])
    Creates and sends a request to the server for [filetype]. This request

>     will not block and will call desired the [callback] upon completion.

Reviewer: ```suggestion
    will not block and will call the desired [callback] upon completion.
```

    Returns the request ID.

    See |lsp#request()| for more information.

    Example:

autocmd CursorHold *
    \ :lua require('vim.lsp').plugin.request_async(
    \     'textDocument/hover',
    \     require('vim.lsp').util.get_text_document_params(),
    \     require('vim.lsp').util.get_filetype()
    \ )

================================================================================
                                                         *lsp-built-in-callback*

The following requests have built-in callbacks defined to handle the response

> in an idiomatic vim fashion.

Reviewer: ```suggestion
in an idiomatic way.
```

  textDocument/definition~
  textDocument/hover~
  textDocument/publishDiagnostics~
  textDocument/references~

require('vim.lsp').config.callbacks.set_all_builtin_callbacks()

>   You can set all built-in callbaks

Reviewer: ```suggestion
  You can enable all built-in callbacks
```

> require('vim.lsp').config..callbacks.set_builtin_callback({method})

Reviewer: ```suggestion
require('vim.lsp').config.callbacks.set_builtin_callback({method})
```

>     You can set a specify built-in callback.

Reviewer: ```suggestion
    You can enable a specific built-in callback.
```

First I understood that this would set/override a builtin/default callback.

Author: It will surely be read like that. :)

        Parameters:~
            {method}: The name of the request, as specified by the
            language server protocol.
                Example: 'textDocument/hover'

vim.lsp.config.callbacks.set_builtin_callbacks({methods})

>     You can set specify built-in callbacks.

Reviewer: ```suggestion
    You can set specific built-in callbacks.
```
s/set/enable ?  (see also above)

        Parameters:~
            {method}: The name of the request, as specified by the
            language server protocol.
                Example: {'textDocument/hover', 'textDocument/publishDiagnostics'}

================================================================================
                                                         *lsp-common-behavior*

> Many of the Language Server Protocol messages are similar to common commandds

Reviewer: ```suggestion
Many of the Language Server Protocol messages are similar to common commands
```

in nvim. Here are some example configurations that could integrate LSP into
your daily workflow.

  textDocument/definition~
    This is essentially "go-to definition". Since this is generally done
    through tags, the following mapping might be considered idiomatic.

    nnoremap <buffer> <C-]> :call lsp#request_async('textDocument/definition')

================================================================================
                                                            *lsp-configuration*

                                                              *lsp-add_check()*
require('lsp.checks').add_check({request}, {cb})
    Configure when methods will be requested.
    Can be used to disable certain methods, if you want the functionality to
    be disabled in the plugin.

    Parameters:~
        {request}: The name of the request, i.e. 'textDocument/hover'

        {cb}: A callback to execute to determine if the {request} should be
        sent.

>         If `false`, the {request} will never be sent.

Reviewer: ```suggestion
        If `false`, the {request} will not be sent.
```

How often / when is this called?  (I assumed it would be done before every (potential) request)

Also could/should document `client` and `request` args for the callback.

    Example:~
        To disable `'textDocument/hover'` in all python files, you could use
        the following snippet in your |init.vim|
>

    lua << EOF
        require('lsp.checks').add_check('textDocument/hover',
            function(client, request)
                if vim.api.nvim_buf_get_option(0, 'filetype') == 'python'
                    return false
                end

                return true
            end
        )
    EOF

<
        To completely disable a message request, you can simply pass

>         `false` as the callback.

Reviewer: ```suggestion
        `false` as the callback, or use the following:
```

>
    :lua vim.lsp.config.callbacks.disable('textDocument/hover')
<

TODO: This should be changed to the Lua API
lsp#configure#file_log_level({level})
    Configure the level of logging to a file.

================================================================================
                                                               *lsp-completion*

To configure completion for a language server, add the following in your

> init.vim

Reviewer: ```suggestion
init.vim:
```

>
  augroup MyLanguageServer
    au!
    autocmd Filetype python inoremap <buffer> <c-n> <c-r>=lsp#completion#complete()<CR>
  augroup END
<

================================================================================
                                                            *lsp-handle-request*

The callbacks used to handle the responses and requests of the language server
protocol are exposed to users of Neovim to allow them to replicate these
actions easily and idiomatically.

For example, if a plugin author wants to replicate the hover functionality of
"textDocument/hover" without wanting to maintain an implementation and cater
to various users configuration preferences, the author of the plugin may use
the Neovim's built-in LSP callbacks to perform the hovering action.

To do so, the following APIs are provided:

>     In VimL, use:  >

Reviewer: ```suggestion
    In Vim script, use:  >
```

                and false means the request failed.

                The parameter {data} is a dictionary containing the possible
                response data (or error) as specified by the language
                server protocol.

                The callback will be executed from Lua running:

                    `vim.api.nvim_call_function('func_name', { success, data })`

            [filetype]: The filetype associated with the server.
            This will default to the current |filetype| if nothing is passed.

            Configuration for the default server can be found at
            |lsp-configuration|.

                                                          *lsp#request_async()*
lsp#request_async({method} [, {arguments} [, {callback} [, {filetype}]]])
    Creates and sends a request to the server for [filetype]. This request
    will not block and will call the desired [callback] upon completion.

    Returns the request ID.

    See |lsp#request()| for more information.

    Example:

autocmd CursorHold *
    \ :lua vim.lsp.request_async(
    \     'textDocument/hover',
    \     vim.lsp.util.get_text_document_params(),

>     \     vim.lsp.util.get_filetype()

Reviewer 2: This argument is not needed.

*lsp.txt* The Language Server Protocol

        NVIM REFERENCE MANUAL

Neovim Language Server Protocol (LSP) API

Neovim exposes a powerful API that conforms to Microsoft's published Language
Server Protocol specification. The documentation can be found here:

	https://microsoft.github.io/language-server-protocol/

================================================================================
                                                            *lsp-server-config*

> Neovim only exposes a powerful API for the language server protocol. To get

Reviewer 3: Sounds weird  to say that "neovim  *only* exposes a powerful API" (if it is powerful, why be modest? ;)) Suggestion: "Neovim exposes a powerful API..., but to get...".

the real benefits of this API, a language server must be installed. Many
examples can be found here:

	https://microsoft.github.io/language-server-protocol/implementors/servers/

After installing a language server to your machine, you must let Neovim know
how to start and interact with that language server. To do so, use the
following APIs:

                                                      *lsp#add_server_config()*
lsp#add_server_config({config}):
	Used to add a language server.

	This will add an autocmd that will start the server and handle any other

> 	startup activities whenever the |FileType| event is triggered for {ftype}.

Reviewer 3: `{ftype}` is not in the signature of the function, better to simply say "for a specified filetype".

	Parameters: ~
		{config} (dictionary): 
			`filetype` (string|array):
				A string or list of strings of filetypes to
				associate with this server.

			`server_name` (string) (optional):
				You can pass any name as server_name.
				If you want to use multiple language servers
				for one filetype, be sure to add server_name.

> 			`cmd` (dictionary): Executing language server command.

Reviewer 2: This is needlessly different from the `jobstart()` syntax. `cmd` should accept a single array where first item is the executable and the rest are args (that luv uses a different syntax internally is irrelevant to the user).

				`execute_path` (string):
					Executing language server command path.
				`args` (array):
					Command arguments.

			`server_config` (dictionary): Optional configuration for a server.

	Example: ~
>
		" C, C++ example
		call lsp#add_server_config({
		      \ 'filetype': ['c', 'cpp'],
		      \ 'server_name': 'cland',

> 		      \ 'cmd': { 'execute_path': 'cland', 'args': ['-background-index'] }})

Reviewer 2: clangd

<
>
		" Go example
		call lsp#add_server_config({
		      \ 'filetype': 'go',
		      \ 'server_name': 'gopls',
		      \ 'cmd': { 'execute_path': 'gopls', 'args': [] }})
<
>
		" Python example
		call lsp#add_server_config({
		      \ 'filetype': 'python',
		      \ 'server_name': 'pyls',
		      \ 'cmd': { 'execute_path': 'pyls', 'args': [] }})
<
>
		" Rust example
		call lsp#add_server_config({
		      \ 'filetype': 'rust',
		      \ 'server_name': 'rls',

> 		      \ 'cmd': { 'execute_path': 'rustup', 'args': ['run', 'stable', 'rls'] },

Reviewer 2: Isn't just `['rls']` better (and use whatever rustup channel that is active for the current environment)?

		      \ 'server_config': {
		      \   'clippy_preference': 'on',
		      \   'all_targets': v:false,
		      \   'build_on_save': v:true,
		      \   'wait_to_build': 0
		      \ }})
<

================================================================================
                                                                  *lsp-request*
Requests

To make a request to the language server client running in Nvim, the following
API is defined:
>
	:lua vim.lsp.request(method, arguments, callback, bufnr, filetype)
	:lua vim.lsp.request_async(method, arguments, callback, bufnr, filetype)
<

                                                            *vim.lsp.request()*
vim.lsp.request({method} [, {arguments} [, {callback} [, {bufnr} [, {filetype} [, {server_name}]]]]]):

> 	Creates and sends a request to the server for [filetype]. Will fill in any

Reviewer 3: [x] to {x} for args in the signature. Here, for `filetype` and below for `callback`.

	missing required arguments as defined by the language server protocol. Will
	call the built-in callback for the request (if you set that and it exists), unless an
	[callback] is specified, in which case that will be called upon completion
	of the request.

	Returns the request's response data.

	Parameters: ~
		{method}: The name of the request, as specified by the
		language server protocol.
			Example: 'textDocument/hover'

		{arguments}: The arguments that should be passed to the language
		server. If any required keys are missing, Nvim will fill those in
		with reasonable default values.
			Example (with 'textDocument/hover'):
				Passed in:
					{} or |v:null|

				Sent:

					{
					    'textDocument': '/home/user/test/example.lua',
					    'position': { 'line': 13, 'character': 17 }
					}

> 		{callback}: An optional callback that will be executed

Reviewer 2: though this API seems over all a little weird. as `vim.lsp.request` returns the result, perhaps the default should be no callback called?

Author: No, `vim.lsp.request` call callback method same as `vim.lsp.request_async` now.
Should we make `vim.lsp.request` not call callback and remove callback from arguments ?

Author: I remove callback argument.
https://github.com/neovim/neovim/pull/10222/commits/181984031d653d93f4706b517cccb66923770258

Author: @bfredl I remove callback argument.
https://github.com/neovim/neovim/pull/10222/commits/181984031d653d93f4706b517cccb66923770258

			once the response has been received. If this is passed and not

> 			|v:null|, then the |lsp-built-in-callback| will not be executed for

Reviewer 2: remove `and not |v:null|`

			this request. If you do not set |built-in-callbacks| or this request does not have a
			specific filetype callback, then no callback will be executed upon receiving the request.

> 			The callback must be a string with the name of the vimscript

Reviewer 2: This sentence can be removed (it is just lua callback now)

			function to call. The signature must be:

				`func_name({success}, {data})`

			where {success} is a boolean where true means successful request
			and false means the request failed.

			The parameter {data} is a dictionary containing the possible
			response data (or error) as specified by the language
			server protocol.

			The callback will be executed from Lua running:

				`vim.api.nvim_call_function('func_name', { success, data })`

		{bufnr}: The number of the buffer.
			This is used for used as an application destination for processing.

		{filetype}: The filetype associated with the server.
			This will default to the current |filetype| if nothing is passed.

		{server_name}: The server name associated with the server.
			This will default to the same value of |filetype|.

> 		Configuration for the default server can be found at |lsp-configuration|.

Reviewer 3: Configuration -> The configuration

                                                      *vim.lsp.request_async()*
vim.lsp.request_async({method} [, {arguments} [, {callback} [, {bufnr} [, {filetype} [, {server_name}]]]]]):

> 	Creates and sends a request to the server for [filetype].

Reviewer 3: [x] to {x}

The following requests have built-in callbacks defined to handle the response
in an idiomatic way.

  textDocument/completion~
  textDocument/hover~
  textDocument/signatureHelp~

>   textDocument/definition~

Reviewer 4: the tilde `~` could have a space before it, so that `*` or `yiw` doesn't include it if user wants to search or yank this word. Though as a side-note, I don't see why `~` highlighting is wanted here.
